### PR TITLE
Feature: object drop hook

### PR DESCRIPTION
**Pull request by iamyulong**
## Summary

Adds `BlueprintHook::OnDrop` and refactors `Proof` blueprints to use this feature.

### For Internal Integrators
Slightly renamed fields in `TypeInfoSubstate` - no structural change.

### CODE REVIEW

File: radix-engine-interface/src/types/blueprint.rs [hunk 1]
```diff
@@ -11,26 +11,24 @@ use scrypto_schema::{InstanceSchema, KeyValueStoreSchema};
...
 pub struct ObjectInfo {
     pub global: bool,
-
     pub blueprint_id: BlueprintId,
-    pub version: BlueprintVersion,
-
-    // Blueprint arguments
-    pub blueprint_info: ObjectBlueprintInfo,
+    pub blueprint_version: BlueprintVersion,
+    pub outer_object: OuterObjectInfo,
```
Comment (talekhinezh) [inline radix-engine-interface/src/types/blueprint.rs:31]: Can you revert these renames? Some of my upcoming PRs refactor `ObjectInfo` a bit.

File: radix-engine/src/errors.rs [hunk 2]
```diff
@@ -150,6 +149,8 @@ pub enum KernelError {
 
     // Invoke
     InvalidInvokeAccess,
+
+    DanglingNodes(Vec<NodeId>),
```
Comment (talekhinezh) [inline radix-engine/src/errors.rs:153]: I quite like the old name better or `OrphanedNodes`

File: radix-engine/src/kernel/actor.rs [hunk 3]
```diff
@@ -9,23 +10,26 @@ pub struct InstanceContext {
...
-    pub module_object_info: ObjectInfo,
+    pub is_direct_access: bool,
+    pub object_info: ObjectInfo,
+}
 
+#[derive(Debug, Clone, PartialEq, Eq, ScryptoSbor)]
+pub struct MethodActor {
     pub ident: String,
+    pub receiver_info: RuntimeReceiverInfo,
+    pub global_address: Option<GlobalAddress>,
```
Comment (talekhinezh) [inline radix-engine/src/kernel/actor.rs:25]: I've also refactored `MethodActor` a bit in https://github.com/radixdlt/radixdlt-scrypto/pull/1246/files#diff-a1773c84b7bd8f3422fd46bd349a608119e9beb86872c8e1f70c1a04ef4d3ecdL14

Could we revert these cleanups?

File: radix-engine/src/system/system_callback.rs [hunk 4]
```diff
@@ -82,6 +86,65 @@ pub struct SystemConfig<C: SystemCallbackObject> {
     pub modules: SystemModuleMixer,
 }
 
+pub fn load_blueprint_definition<Y, C>(
```
Comment (talekhinezh) [inline radix-engine/src/system/system_callback.rs:89]: Feels to me that this should still be in `system.rs` as it's part of the "service" the system provides. Is there a reason to move it here?

File: radix-engine/src/kernel/actor.rs (hunk 3 above)
Comment (iamyulong) [reply radix-engine/src/kernel/actor.rs:25]: Briefly looked at the other PR. I'm not sure I agree with that refactor.

I'm happy to revert any specific change here, if it doesn't look right to you.

File: radix-engine/src/system/system_callback.rs (hunk 4 above)
Comment (iamyulong) [reply radix-engine/src/system/system_callback.rs:93]: It was to make it usable by both `system.rs` and `system_call.rs`, as it's now purely a utility function.

Happy to move to `SystemService` (seems like another centralised place for such functionalities).

Comment (iamyulong) [reply radix-engine/src/system/system_callback.rs]: Moved to `SystemService`.

All I need to do is to construct a `SystemService::new(api)`. No idea why I didn't do that initially.

File: radix-engine/src/errors.rs (hunk 2 above)
Comment (iamyulong) [reply radix-engine/src/errors.rs:153]: Yes, will change.

File: radix-engine/src/kernel/actor.rs (hunk 3 above)
Comment (talekhinezh) [reply radix-engine/src/kernel/actor.rs:25]: Yup, I think just removing `RuntimeReceiverInfo` and having all info flat in `MethodActor` for now would make it easiest. For example, I consider `object_info` as not required but is essentially just a cache of this info for `MethodActor`.

**Review by talekhinezh — APPROVED** [review]

File: radix-engine/src/kernel/actor.rs (hunk 3 above)
Comment (iamyulong) [reply radix-engine/src/kernel/actor.rs:21]: Yes, removed it.

File: radix-engine-interface/src/types/blueprint.rs (hunk 1 above)
Comment (iamyulong) [reply radix-engine-interface/src/types/blueprint.rs:34]: Reverted `blueprint_version`, but kept `outer_object` as I feel it's more accurate.

**Pull request merged**